flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunctionTest.java

```diff
@@ -35,60 +42,101 @@
 import java.io.IOException;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
+import java.time.Clock;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.hasItem;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 /**
  * Tests for {@link TwoPhaseCommitSinkFunction}.
  */
 public class TwoPhaseCommitSinkFunctionTest {
-	TestContext context;
+
+	@Rule
+	public TemporaryFolder folder = new TemporaryFolder();
+
+	private FileBasedSinkFunction sinkFunction;
+
+	private OneInputStreamOperatorTestHarness<String, Object> harness;
+
+	private AtomicBoolean throwException = new AtomicBoolean();
+
+	private File targetDirectory;
+
+	private File tmpDirectory;
+
+	@Mock
+	private Clock mockClock;
+
+	@Mock
+	private Logger mockLogger;
```

Comment:
As above. Will all of those `mockLogger` tests fail, if you replace:
```
			log.warn("Transaction {} has been open for {} ms. " +
					"This is close to or even exceeding the transaction timeout of {} ms.",
				transactionHolder.handle,
				elapsedTime,
				transactionTimeout);
```
with 
```
			log.warn("Transaction {} has been open for too long. " +
					"This is close to or even exceeding the transaction timeout of {} ms.",
				transactionHolder.handle,
				transactionTimeout);
```
or
```
			log.warn(String.format("Transaction {} has been open for {} ms. " +
					"This is close to or even exceeding the transaction timeout of {} ms.",
				transactionHolder.handle,
				elapsedTime,
				transactionTimeout));
```
?
If so, then this is a perfect example my I consider mockito to be the definition of evil. Using mockito in tests in 99% cases is duplicating/copying productional code into tests, basically repeating the implementation, which is super fragile and brakes on even the tiniest refactors. While good tests instead should check/assert the actual effects - not whether specific method calls were called and how many times they were called.

Reply:
The test is asserting on the presence of the substring `This is close to or even exceeding the transaction timeout` in the log message. All your changes to the code would still pass the test except for the 2nd case because there is an assert on the elapsed time. Mockito is used here so that the argument passed to `.warn` can be captured. Imo this is not evil as no behavior is mocked, and actual effects are tested.

Reply:
Are you sure? `Logger.warn(String, Object...)`, `Logger.warn(String, Object)`, `Logger.warn(String, Object, Object)` and `Logger.warn(String)` are different methods.

This shows, that you are not testing for the effects (warning message being logged somewhere), but you test whether one particular method was called or not.

Reply:
True, you are right. One would need to implement the `org.slf4j.Logger` interface.
- http://projects.lidalia.org.uk/slf4j-test/ provides a `org.slf4j.Logger` implementation for unit testing but this approach comes with other problems (e.g. https://github.com/Mahoney/slf4j-test/issues/15, [NOP if log level is  disabled](https://github.com/Mahoney/slf4j-test/blob/master/src/main/java/uk/org/lidalia/slf4jtest/TestLogger.java#L449)

Right now I don't see a way to do it properly. Any help is welcome.

Reply:
The test does not rely on any mocks anymore. I am not 100% happy with it because we use log4j `1.x` which is not maintained anymore and in log4j `2.x`, the APIs have changed a lot: http://logging.apache.org/log4j/2.x/manual/customconfig.html#AddingToCurrent